src/dumper_thread.c

```diff
@@ -550,6 +550,19 @@ cr_dumper_thread(gpointer data, gpointer user_data)
     }
 #endif
 
+    // Allow checking that the same package (NEVRA) isn't present multiple times in the metadata
+    // Keep a hashtable of NEVRA mapped to an array-list of location_href values
+    g_mutex_lock(&(udata->mutex_nevra_table));
+    gchar *nevra = cr_package_nevra(pkg);
+    GArray *pkg_locations = g_hash_table_lookup(udata->nevra_table, nevra);
+    if (!pkg_locations) {
+        pkg_locations = g_array_new(FALSE, TRUE, sizeof(gchar *));
+        g_hash_table_insert(udata->nevra_table, nevra, pkg_locations);
+    }
+    gchar *location = g_strdup(pkg->location_href);
+    g_array_append_val(pkg_locations, location);
+    g_mutex_unlock(&(udata->mutex_nevra_table));
```

Comment:
It looks pretty innocent, (O(logN) I hope.  But actually we might need to opt out even the calculating of this hash.  In Copr with the [giant repositories like this](https://download.copr.fedorainfracloud.org/results/@rubygems/rubygems/fedora-rawhide-x86_64/repodata/), we use ` --no-stat --update --recycle-pkglist` to speedup metadata updates, which in turn means that createrepo_c isn't doing fancy with the `pkg` structure.  Now we use it to calculate the hash.

This deserves some benchmark IMO.

Reply:
I'm running this command currently with the nonmodified `createrepo_c` and it is taking upwards of 10 minutes, is that expected?

Reply:
Yeah, so, this is why.  85% of the runtime is spent inside of `g_queue_insert_sorted()`  That is crazy.  Given this, I'm not sure measuring any additional impact of my code is a great use of time. I don't expect it to even register, given that producing and compressing multiple gigabytes of metadata just barely registers as a tiny blip way over on the right-hand side of the chart.

![flamegraph](https://user-images.githubusercontent.com/2751996/175654622-37b97aa5-fc60-48d5-9483-94a24a61735e.svg)

Reply:
Filed as https://github.com/rpm-software-management/createrepo_c/issues/323

Reply:
Nice analysis, @dralley!